ICSharpCode.Decompiler/IL/ILReader.cs

```diff
@@ -575,10 +575,26 @@ void InsertStackAdjustments()
 		/// <summary>
 		/// Decodes the specified method body and returns an ILFunction.
 		/// </summary>
-		public ILFunction ReadIL(MethodDefinitionHandle method, MethodBodyBlock body, GenericContext genericContext = default, ILFunctionKind kind = ILFunctionKind.TopLevelFunction, CancellationToken cancellationToken = default)
+		public ILFunction ReadIL(MethodDefinitionHandle method, MethodBodyBlock body, GenericContext genericContext = default, 
+			ILFunctionKind kind = ILFunctionKind.TopLevelFunction, CancellationToken cancellationToken = default,
+			IncrementalChecksum checksumcalc = null)
 		{
 			cancellationToken.ThrowIfCancellationRequested();
 			Init(method, body, genericContext);
+
+			if (checksumcalc != null && checksumcalc.Enabled)
```

Comment:
This should not be part of ILReader! Why are you adding unrelated functionality in random places?

Reply:
I need to know function name, so I can checksum it. Init in that phase determines function name, after that further decompilation is not needed - that's why breaking there.

But my test coverage is rather low right now, need to check deeper what kind of changes there are expected and what and where to checksum.

Do you know correct place where strings reside ?
https://github.com/tapika/ILSpy/blob/master/ICSharpCode.Decompiler.Tests/ChecksumTest.cs#L78

this test succeeds, but it should fail. Probably need to checksum data as well.

Reply:
I suggest you take a look at how .NET assemblies are structured (see ECMA-335 https://www.ecma-international.org/publications-and-standards/standards/ecma-335/). The .NET metadata is generally organized in tables, except for strings, blobs and GUIDs, which are stored in a separate stream (called heap). Also, the IL code is stored in another section of the executable.

There are tables for type definitions, type references, method definitions, etc. Each method/type defined in an assembly has a row in the respective table. The MethodDef table for example has a column for the method name.

What ILSpy does is the following: it uses the SRM PEReader class to read the PE file (You can use this nuget package provided by Microsoft: https://www.nuget.org/packages/System.Reflection.Metadata) and then instantiates a MetadataReader, which then can be used to browse all metadata tables. For example:

```csharp
PEReader reader = ...;
MetadataReader metadata = reader.GetMetdataReader();

MethodDefinition method = metadata.GetMethodDefinition(MetadataTokens.MethodDefinitionHandle(1)); // this retrieves the data of the "first" method defined in the assembly.
string methodName = metadata.GetString(method.Name); // retrieves the name of the first method as string...
```

If you want to look at the structure of an assembly, you can use the metadata explorer included in ILSpy (beneath the assembly tree node, there is the "Metadata" node, which offers access to all metadata tables in the assembly).

I hope this helps!

Reply:
It's indeed much easier to get into magical world of ECMA-335 specs without ECMA-335 specs.

I've checked through - documentation itself is ~ 500 pages long , and finding right information seems to be almost impossible.

With ilspy I have in several rounds get into point what I want to check in assembly and how it works.

Strings which I'm trying to find at the moment are called user strings, and ilspy have code snipet
`return metadata.GetUserString(MetadataTokens.UserStringHandle(blob.ReadInt32()));`

which obtains that string.

But if I want to enumerate all user strings - I was not able to find any code, which can perform something like this.

Even in test application `metadata.GetUserString` works with `UserStringHandle.FromOffset(1)` and `UserStringHandle.FromOffset(0x1B)` - so suspect strings are just serialized one after another without being able to enumerate them.

Concerning long term development strategy - what if I'll split my changes into smaller parts and then make changes in step by step manner. Maybe hacking ILReader is not best approach after all, but I need to identify best strategy of adding check summing functionality into ilspy.

Reply:
From the spec:
![image](https://user-images.githubusercontent.com/344204/111114405-79674680-8563-11eb-94b3-523f27febd05.png)

Reply:
So, you will have to either scan all method bodies for ldstr instructions and inspect their referenced user strings, or scan the heap and filter out the garbage, but there is no table for user strings, which could be read in a fast way.

Take a look at https://github.com/icsharpcode/ILSpy/blob/e08057a4a4d7f60458925765d7e06223c8ca70e0/ILSpy/Search/LiteralSearchStrategy.cs for an example on how to look for strings/user-strings. Note that the code handles all instructions individually, but you could also use ILParser.GetOperandType to simplify the code.

Hope this helps!

Reply:
Btw - it's also possible to use

https://github.com/0xCM/z0/blob/fbeab69df16d550c88e65d3e402a6d62d942e068/src/cli/src/pe/reader/UserStrings.cs

`reader.GetNextHandle(handle);`

will try this later on maybe.